plugins/modules/purefa_pg.py

```diff
@@ -379,110 +412,187 @@ def make_pgroup(module, array):
+                    names=[module.params["name"]],
+                    protection_group=ProtectionGroup(retention_lock="ratcheted"),
                 )
-            except Exception:
+            if res.status_code != 200:
                 module.fail_json(
-                    msg="Failed to set SafeMode on pgroup {0}".format(
-                        module.params["name"]
+                    msg="Failed to set SafeMode on pgroup {0}. Error: {1}".format(
+                        module.params["name"], res.errors[0].message
                     )
                 )
     module.exit_json(changed=changed)
 
 
 def rename_exists(module, array):
     """Determine if rename target already exists"""
-    exists = False
+    api_version = array.get_rest_version()
     new_name = module.params["rename"]
     if ":" in module.params["name"]:
         container = module.params["name"].split(":")[0]
         new_name = container + ":" + module.params["rename"]
         if "::" in module.params["name"]:
             new_name = container + "::" + module.params["rename"]
-    for pgroup in array.list_pgroups(pending=True):
-        if pgroup["name"].casefold() == new_name.casefold():
-            exists = True
-            break
-    return exists
+    if LooseVersion(CONTEXT_API_VERSION) <= LooseVersion(api_version):
+        res = array.get_protection_groups(
+            names=[new_name], context_names=[module.params["context"]]
+        )
+    else:
+        res = array.get_protection_groups(names=[new_name])
+    if res.status_code == 200:
+        return True
+    return False
 
 
 def update_pgroup(module, array):
     """Update Protection Group"""
+    api_version = array.get_rest_version()
     changed = renamed = False
     state = module.params["state"]
-    api_version = array._list_available_rest_versions()
     if module.params["target"]:
         connected_targets = []
-        connected_arrays = get_arrays(array)
-
-        if OFFLOAD_API_VERSION in api_version:
-            connected_targets = get_targets(array)
+        connected_arrays = get_arrays(module, array)
+        connected_targets = get_targets(module, array)
         connected_arrays = connected_arrays + connected_targets
         if connected_arrays == []:
             module.fail_json(msg="No targets connected to source array.")
-        current_connects = array.get_pgroup(module.params["name"])["targets"]
+        if LooseVersion(CONTEXT_API_VERSION) <= LooseVersion(api_version):
+            current_connects = array.get_protection_groups_targets(
+                group_names=[module.params["name"]],
+                context_names=[module.params["context"]],
+            )
+        else:
+            current_connects = array.get_protection_groups_targets(
+                group_names=[module.params["name"]]
+            )
         current_targets = []
 
         if current_connects:
             for targetcnt in range(0, len(current_connects)):
```

Comment:
current_connects is a direct ValidResponse object, should do necessary conversion before using.